js/diff.js

```diff
@@ -0,0 +1,116 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+function calcDiffList(viewNode, prevNode) {
+    const PROPERTY_LIST = [
+        "classnameIndex",
+        "id",
+        "left",
+        "top",
+        "width",
+        "height",
+        "scrollX",
+        "scrollY",
+        "translationX",
+        "translationY",
+        "scaleX",
+        "scaleY",
+        "alpha",
+        "willNotDraw",
+        "clipChildren",
+        "visibility",
+        "elevation"
+    ]
+
+    class Diff {
+        constructor() {
+            this.newNode = null
+            this.newChildren = []
+            this.removedChildren = []
+            this.changedProperties = []
+            this.updateBoxPos = false
+            this.hasChildOrderChanged == false
+        }
+    
+        isEmpty() {
+            return this.newChildren.length == 0
+                && this.removedChildren.length == 0
+                && this.changedProperties.length == 0
+                && this.boxStylePos != null
+                && !this.hasChildOrderChanged
+        }
+    }   
+
+    function computeDiffInner(newNode, oldNode) {
+        // check if the nodes have the same attributes. Make Notes of any differences
+        const diff = new Diff()
+        diff.newNode = newNode
+
+        for (let pName of PROPERTY_LIST) {
```

Comment:
you only care about the property diff of the selected node, no need to parse diff of every node here, just compare the selected node during rendering

Reply:
The reason I am doing this comparison is because el.node references aren't being updated if there is no difference between frames. This means that we need to update the properties when the user switches frames, or otherwise find another way to ensure that when the user selects a node the correct property values show up.

I also considered using more global state, such as a hashmap of node Id -> Node[] of size "frameCount", then use that whenever the user selects the frame. It would take constant lookup time, but would also require more memory usage. This approach also resulted in a really ugly mixture of certain parts of the class using the DOM's el.node references, and some parts of the code using the hashmap object to get the correct node.

I also considered a linear search lookup whenever the user clicked on a node. That route is actually seeming more appealing now that I consider it more.

What are your thoughts?

Reply:
The most recent revision treats the rootNodes[] array as the source of truth, and simply updates the node references in the comparing logic, like we talked about during our 1:1. Doing this allows us to remove the "document.getElementbyId(node.treeDisplayName) lookups"